client/worker.go

```diff
@@ -0,0 +1,241 @@
+				err.Permanent = true
+				return err
+			}
+			time.Sleep(time.Duration(err.RetryDelaySeconds) * time.Second)
+			retries++
+			continue
+		}
+		return err
+	}
+	return nil
+}
+
+func (w *worker) send(ctx context.Context, r *sendRequest) *ErrHTTPSend {
+	_, span := trace.StartSpan(ctx, "export")
+	defer span.End()
+
+	req, err := http.NewRequest("POST", w.endpoint, bytes.NewBuffer(r.message))
+	if err != nil {
+		span.SetStatus(trace.Status{
+			Code:    trace.StatusCodeInvalidArgument,
+			Message: err.Error(),
+		})
+		return &ErrHTTPSend{Err: err, Permanent: true}
+	}
+	req.Header.Add(headerContentType, headerValueXProtobuf)
+	req.Header.Add(headerContentEncoding, headerValueGZIP)
+	if w.authToken != "" {
+		req.Header.Add(headerAuthToken, w.authToken)
+	}
+
+	resp, err := w.client.Do(req)
+	if err != nil {
+		span.SetStatus(trace.Status{
+			Code:    trace.StatusCodeInternal,
+			Message: err.Error(),
+		})
+		return &ErrHTTPSend{Err: err}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+		return nil
+	}
+
+	// Drop the batch if server thinks it is malformed in some way or client is not authorized
+	if resp.StatusCode == http.StatusBadRequest || resp.StatusCode == http.StatusUnauthorized {
+		msg := fmt.Sprintf("server responded with: %d", resp.StatusCode)
+		span.SetStatus(trace.Status{
+			Code:    ocTrancelator.OCStatusCodeFromHTTP(int32(resp.StatusCode)),
+			Message: msg,
+		})
+		return &ErrHTTPSend{
+			Err:        fmt.Errorf("dropping request: %s", msg),
+			StatusCode: http.StatusBadRequest,
+			Permanent:  true,
+		}
+	}
+
+	// Check if server is overwhelmed and requested to pause sending for a while.
+	// Pause this worker from sending more data till the specified number of senconds in the Retry-After haeder
+	// Fallback to defaultRateLimitingBackoffSeconds if the header is not present
+	if resp.StatusCode == http.StatusTooManyRequests {
+		retryAfter := defaultRateLimitingBackoffSeconds
+		if val := resp.Header.Get(headerRetryAfter); val != "" {
+			if seconds, err := strconv.Atoi(val); err == nil {
+				retryAfter = seconds
+			}
+		}
+		span.SetStatus(trace.Status{Code: trace.StatusCodeResourceExhausted})
+		return &ErrHTTPSend{
+			Err:               errors.New("server responded with 429"),
+			StatusCode:        resp.StatusCode,
+			RetryDelaySeconds: retryAfter,
+		}
+	}
+
+	// TODO: handle 301, 307, 308
+	// redirects are not handled right now but should be to confirm with the spec.
+
+	span.SetStatus(trace.Status{Code: ocTrancelator.OCStatusCodeFromHTTP(int32(resp.StatusCode))})
+	return &ErrHTTPSend{
+		Err:        fmt.Errorf("error exporting spans. server responded with status %d", resp.StatusCode),
+		StatusCode: resp.StatusCode,
+	}
+}
+
+// prepare takes a jaeger batch, converts it to a SAPM PostSpansRequest, compresses it and returns a request ready
+// to be sent. The method is not safe to be called from multiple goroutines. Each caller must use locks to avoid races
```

Comment:
Unclear regarding safety. Does this only use worker fields? What are the shared fields that make this unsafe? Isn't this called only from its own worker? I do not see any locks at the call site.

Reply:
The shared gzip writer is mainly what is shared. You're right that worker itself is not locking anything and we expect the caller (client) to ensure it exports synchronously from a single worker. Contractually that makes the whole worker unsafe for concurrent ops and caller should use locks. I'll update the comment.